src/Data/Array/Destination.hs

```diff
@@ -2,9 +2,86 @@
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
+-- | This module provides destination arrays
+--
+-- We can use linear types to release a safe API for destination arrays
+-- and DPS (destination passing style).
+--
+-- == What are destination arrays and DPS?
+--
+-- There are several lists which must, despite program optimizations
+-- (i.e., [deforesting](https://www.sciencedirect.com/science/article/pii/030439759090147A)),
+-- be allocated, filled, passed along and de-allocated. When the allocation
+-- of these arrays is controlled by the programmer and not
+-- done by Haskell's GC (garbage collector), programs are often more efficient.
+--
+-- DPS is a style of programming in which functions that produce such arrays
+-- take them as arguments -- moving the responsibility of allocating, using
+-- and freeing those arrays to the caller. The arrays passed as arguments
+-- to functions that must be written to are called /destination arrays/.
+--
+-- Here is a C function that adds vectors written in DPS style; it
+-- takes in the destiniation array @sum@ and writes to it:
+--
+-- @
+-- void add(int size, int *a, int *b, int *sum){
+--   for (int i=0; i<size;++i){sum[i]=a[i]+b[i];}
+-- }
+-- @
+--
+-- == Linear Types Enforce Destination Arrays are Filled
+--
+-- Linear types are used to ensure that the destination array
+-- is always written to. Why? Well:
+--
+-- Because of linear types, any function that uses a destination array
```

Comment:
"because of linear types" sounds a bit hand-wavy. The explanation after this sentence is clear, so maybe we should just start with that ("The only way to create ...").